sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -451,5 +452,24 @@ case class StringToMap(text: Expression, pairDelim: Expression, keyValueDelim: E
     mapBuilder.build()
   }
 
+  override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val builderTerm = ctx.addReferenceObj("mapBuilder", mapBuilder)
+    val keyValues = ctx.freshName("kvs")
+    val idx = ctx.freshName("i")
+
+    nullSafeCodeGen(ctx, ev, (text, pd, kvd) =>
+      s"""
+         |int $idx = 0;
+         |UTF8String[] $keyValues = $text.split($pd, -1);
+         |while ($idx < $keyValues.length) {
```

Comment:
shall we use for loop? then we don't need to declare `int idx` outside of the loop.

Reply:
make `idx` as `freshName` because  https://github.com/apache/spark/blob/c4115793555d35a10c79038ce51dc5f803f3b838/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateMutableProjection.scala#L132 has defined `i` in scope, seems only happen in test cases?